ptychonn/_infer/__main__.py

```diff
@@ -0,0 +1,202 @@
+    ),
+)
+@click.argument(
+    'out_dir',
+    type=click.Path(
+        exists=True,
+        path_type=pathlib.Path,
+    ),
+)
+def infer_cli(
+    data_path: pathlib.Path,
+    scan_path: pathlib.Path,
+    out_dir: pathlib.Path,
+):
+    '''Infer a reconstructed image from diffraction patterns at DATA_PATH and
+    scan positions at SCAN_PATH. Save inferred patches and stitched image at
+    OUT_DIR.
+    '''
+
+    inferences_out_file = out_dir / 'inferences_506.npz'
+    click.echo(f'Does data path exist? {data_path.exists()}')
+
+    with h5py.File(data_path) as f:
+        data = f['entry/data/data'][()]
+
+    inferences = infer(
+        data=data,
+        inferences_out_file=inferences_out_file,
+    )
+
+    ## parameters required for stitching individual inferences
+    spiral_step = 0.05
+    step = spiral_step * -1e-6
+    spiral_traj = np.load(scan_path)
+    scan = np.stack((spiral_traj['x'], spiral_traj['y']), axis=-1) * step
+    stitched = stitch_from_inference(
+        inferences,
+        scan,
+        stitched_pixel_width=10e-9,
+        inference_pixel_width=10e-9,
+    )
+
+    # Plotting some summary images
+    plt.figure(1, figsize=[8.5, 7])
+    plt.pcolormesh(stitched)
+    plt.colorbar()
+    plt.tight_layout()
+    plt.title('stitched_inferences')
+    plt.savefig(out_dir / 'stitched_506.png', bbox_inches='tight')
+
+    test_inferences = [0, 1, 2, 3]
+    fig, axs = plt.subplots(1, 4, figsize=[13, 3])
+    for ix, inf in enumerate(test_inferences):
+        plt.subplot(1, 4, ix + 1)
+        plt.pcolormesh(inferences[inf])
+        plt.colorbar()
+        plt.title('Inference at position {0}'.format(inf))
+    plt.tight_layout()
+    plt.savefig(out_dir / 'inferences_0_to_4_scan506.png', bbox_inches='tight')
+
+    return 0
+
+
+def infer(
+    data: npt.NDArray,
+    *,
+    inferences_out_file: typing.Optional[pathlib.Path] = None,
+) -> npt.NDArray:
+    '''Infer ptychography reconstruction for the given data.
+
+    For each diffraction pattern in data, the corresponding patch of object is
+    reconstructed.
+
+    Set the CUDA_VISIBLE_DEVICES environment variable to control which GPUs
+    will be used.
+
+    Parameters
+    ----------
+    data : (POSITION, WIDTH, HEIGHT)
+        Diffraction patterns to be reconstructed.
+    inferences_out_file : pathlib.Path
+        Optional file to save reconstructed patches.
+
+    Returns
+    -------
+    inferences : (POSITION, WIDTH, HEIGHT)
+        The reconstructed patches inferred by the model.
+    '''
+
+    with importlib.resources.path(
+            'ptychonn._infer',
+            'weights.pth',
+    ) as load_model_path:
+        click.echo(f'Loading model at {load_model_path}')
+        recon_model = helper_small_model.ReconSmallPhaseModel()
+        tester = helper_small_model.Tester(
+            model=recon_model,
+            batch_size=max(torch.cuda.device_count(), 1) * 64,
+            model_params_path=load_model_path,
+        )
```

Comment:
@stevehenke requested that the ptychonn.infer() API allow arbitrary model inputs, but remain a datastructure (not file-based API), so I need to move the instantiation of this "Tester" class outside of this function.